### ISSUE TITLE
Pocket PLA update wPocket using test data set?

### ISSUE THREAD

**Issue by digdig** [hsiaoyi0504/ML2015#1]
I have a question about machine learning, I hope you don't mind me opening this issue.

In hw1, problem 18, when update `pocketW`, you use the `test data` rather than `training data`. I thought it might be more proper to use the `training data` throughout the learning process. After we get the **final** `pocketW`, we might using the `testing data` to verify how well we have learned using pocket algorithm.

I want to hear your thought on why you use the `training data` to find the bad points and update `w`, but decide whether to accept this update by  the `testing data`(in my opinion this should also use the training data)?

``` python
if (innerProduct<=0)!= (y[randomNum]<0):
            for k in range(dimensions): #here you are using training data
                w[k]=w[k]+updateFactor*y[randomNum]*x[k][randomNum] 
            updates=updates+1
            #check if update pocketW
            testErrorCount=0
            for j2 in range(len(x2[0])): #here you are using testing data
                innerProduct=0;
                for k in range(dimensions):
                    innerProduct=innerProduct+w[k]*x2[k][j2]
                if (innerProduct<=0)!= (y2[j2]<0):
                    testErrorCount+=1
            if testErrorCount<pocketErrorTimes:
                pocketW=w
                pocketErrorTimes=testErrorCount

```

**Comment by hsiaoyi0504** [comment]
I feel surprised that someone asks a question here XD

I didn't really get it, but it seems that you mean why I update weight based on testing rather than training data? After you learn more in the future, you will know it's due to generalization. You can train a model perfectly fitting your training data, but in the real world, we usually want to predict out-of-example cases (something never happen in the training set). Consequently, we prefer generate a general model rather than only performing well on training set. Then, we probe this through validation( like the testing set here). They are some theory behind this actually, something related to statistics.

**Comment by digdig** [comment]
![screenshot_2016-12-18_01-20-21](https://cloud.githubusercontent.com/assets/5459093/21288519/4604aff2-c4c0-11e6-9112-560d99e54037.png)

In the lecture notes(especially the formula there), I think the training is based on the minimization of the `bad points` in the training sets. After all, when apply a pocket PLA algorithm, no testing data is needed in prior I think. **Training is training, testing is testing in my opinion.**

You can test the training result using the testing data, if you update the weight using the testing data, I think you are considering the testing data as part of the training data as well. 

I have a backup post here, it didn't use the testing data to update. http://blog.csdn.net/a1015553840/article/details/50979640

yes, I am still learning. finding your repo is very helpful for me.

**Comment by digdig** [comment]
> You can train a model perfectly fitting your training data, but in the real world, we usually want to predict out-of-example cases (something never happen in the training set).

I don't think it really apply to this problem, because your training data is linear non-separable. You are not going to get a perfectly weight in the first place, that's the reason we are using pocket PLA.

**Comment by hsiaoyi0504** [comment]
I think you can take a look at "Lecture 5: Training versus Testing" first.
BTW, what I mentioned is just a case. What I mean here is that you can do very well on training set, but it doesn't imply that you will do well on any data. Especially, if you train your model only through training data, you will get very good performance and you usually will not perform well on out-of-example data (overfitting). In fact, select model based on testing (validation exactly) promises you that you will get approximately well on training data and out-of-example data.

**Comment by hsiaoyi0504** [comment]
After few classes latter, you will learn that we usually do only have a dataset, but we will divide it into two parts, training set and validation set (testing data in this case). Then, we train our model based on training set, but we determine if we should use updated model is based on the performance on validation set.

**Comment by hsiaoyi0504** [comment]
Maybe this [link](https://www.quora.com/What-is-a-training-data-set-test-data-set-in-machine-learning-What-are-the-rules-for-selecting-them) can help you.

**Comment by hsiaoyi0504** [comment]
> You can test the training result using the testing data, if you update the weight using the testing data, I think you are considering the testing data as part of the training data as well.

Yes, it is what I mean. Are you taking the physical classes or online classes? For homework 1, I remembered that the teacher didn't introduce validation at that time, and he introduced that later. In fact, the name "testing data" here is somewhat confusing, but in this homework, it does mean validation (supervise the trained model or check if we should use updated weights)

**Comment by digdig** [comment]
Thanks for the link, it is really helpful to know that the test data in this case is actually the validation set, and we actually *tune the parameter using the validation set* .  Your clarification is also very helpful, I will pay attention to the points you mentioned in my future learning.

**Comment by digdig** [comment]
I am taking the online class, it is really high standard.

**Comment by digdig** [comment]
I think things are clear to me now, I really appreciate your clarification.

**Comment by hsiaoyi0504** [comment]
You are welcome. I think discussion make things clear and I like discussion, but I worried that I can't be helpful because it was the homework one year ago. XD
